### ISSUE TITLE
Add distributed examples into PyTorch CI tests

### ISSUE THREAD

**Issue by mrshenli** [pytorch/pytorch#46321]
### Targeted Examples

This will help us to capture tutorial/examples failures caused by BC-breaking changes. So the test should ideally be the same as the examples. However, as some examples depends on non-PyTorch libraries, we will need to replace those with some dummy implementation. Below is a list of existing examples.

- [ ] [Distributed Reinforcement Learning using RPC and RRef](https://pytorch.org/tutorials/intermediate/rpc_tutorial.html#distributed-reinforcement-learning-using-rpc-and-rref). Example source code is [here](https://github.com/pytorch/examples/tree/master/distributed/rpc/rl). One caveat is that, since we cannot let PyTorch depend on `gym`, we will need to create a dummy environment. This is fine, as we are only testing API correctness. It should be sufficient to return random states and let the training loop to run a small number of iterations.
- [ ] [Distributed RNN using Distributed Autograd and Distributed Optimizer](https://pytorch.org/tutorials/intermediate/rpc_tutorial.html#distributed-rnn-using-distributed-autograd-and-distributed-optimizer). Example source code is [here](https://github.com/pytorch/examples/tree/master/distributed/rpc/rnn).
- [ ] [Implementing a Parameter Server using Distributed RPC Framework](https://pytorch.org/tutorials/intermediate/rpc_param_server_tutorial.html). Example source code is [here](https://github.com/pytorch/examples/tree/master/distributed/rpc/parameter_server).
- [ ] [Batch-Updating Parameter Server](https://pytorch.org/tutorials/intermediate/rpc_async_execution.html#batch-updating-parameter-server). Source code is [here](https://github.com/pytorch/examples/blob/master/distributed/rpc/batch/parameter_server.py). Similar to the first item, we cannot make PyTorch depend on torchvision, so please choose some dummy models instead.
- [ ] [Batch-Processing CartPole Solver](https://pytorch.org/tutorials/intermediate/rpc_async_execution.html#batch-processing-cartpole-solver). Source code is [here](https://github.com/pytorch/examples/blob/master/distributed/rpc/batch/reinforce.py). The test cannot depend on `gym`, so please use some dummy environments.
- [ ] [Distributed Pipeline Parallelism Using RPC](https://pytorch.org/tutorials/intermediate/dist_pipeline_parallel_tutorial.html). Source code is [here](https://github.com/pytorch/examples/tree/master/distributed/rpc/pipeline). Again, cannot depend on `torchvision`.
- [ ] [Combining DDP with RPC](https://pytorch.org/tutorials/advanced/rpc_ddp_tutorial.html). Source code is [here](https://github.com/pytorch/examples/tree/master/distributed/rpc/ddp_rpc).

### Implementation

TBA

**Comment by pritamdamania87** [comment]
@mrshenli It looks like we're adding a copy of our examples into the PyTorch repo (https://github.com/pytorch/pytorch/pull/46510). I thought the examples repo itself has a CI and we had a plan to extend the CI on that repo to run the examples? It would be hard to maintain two copies of the same source.

**Comment by mrshenli** [comment]
@pritamdamania87 yep, the best option would be doing that on the CI, if the example CI can kick off a job for every PyTorch PR. Otherwise, we will only get delayed signals and will have to bisect to find the culprit PRs.
